libbeat/outputs/kafka/client.go

```diff
@@ -220,6 +276,13 @@ func (c *client) errorWorker(ch <-chan *sarama.ProducerError) {
 	}
 }
 
+func (c *client) retry(r *msgRef) {
+	go func() {
+		r.backoff.Wait()
+		publish <- r
+	}()
```

Comment:
This is kind of an anti-pattern. 

It seems we start the go-routine to not create a deadlock between event publishing end error/success reporting.

The effect of starting a go-routine for every 'message' is similar to implementing an unbounded queue. Just that a go-routine requires 2KB initial stack size.

The go-routine also waits and blocks the re-scheduling msgRef. Only after this period will the msg tried to be forwarded. In the meantime the beats internal queue will try to publish more batches (as Publish is not blocking).

The `retry` flag is always set to true, forcing the kafka client to shutdown and restart completely. Problem is:
- we still try to publish batches while we backoff
- we enforce a full shutdown and reconnect to cluster, even if the client lib has been recovered itself in the meantime

The memory queue (default queue implementation) has an upper bound of active events in the beat. At some point all events might have reached the kafka output, because all batches might end up in separate retry go-routines at worst. This is kinda ok for the memory queue, but once the spool file is configure, millions of events might hit the output at worst case, resulting in us OOMing. The spool does not enforce an upper limit of active events, such that we require the outputs to block if they are unstable.

Sending the `batch.Retry()` signal can also block the beats internal queues, such that no more events are written from the spool.